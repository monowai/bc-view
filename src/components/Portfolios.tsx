import React, { ReactElement, useState } from "react"
import { useTranslation } from "next-i18next"
import useSwr from "swr"
import { portfoliosKey, simpleFetcher } from "@utils/api/fetchHelper"
import { Portfolio } from "types/beancounter"
import { useRouter } from "next/router"
import { rootLoader } from "@components/PageLoader"

export function Portfolios(selectedPortfolio: Portfolio): ReactElement {
  const { data } = useSwr(portfoliosKey, simpleFetcher(portfoliosKey))
  const { t, ready } = useTranslation("common")
  const router = useRouter()
  const [isOpen, setIsOpen] = useState(false)
  const [selected, setSelected] = useState(selectedPortfolio)

  if (!ready || !data) {
    return rootLoader(t("loading"))
  }
  const portfolios: Portfolio[] = data.data

  const handleSelect = (portfolio: Portfolio): void => {
    setSelected(portfolio)
    setIsOpen(false)
    router.push(`${portfolio.code}`).then()
  }

  return (
    <div className="relative inline-block text-left">
      <div>
        <button
          type="button"
          className="inline-flex justify-center w-full rounded-md border border-gray-300 shadow-sm px-4 py-2 bg-white text-sm font-medium text-gray-700 hover:bg-gray-50 focus:outline-none"
          onClick={() => setIsOpen(!isOpen)}
          style={{ zIndex: 10 }} // Set a lower z-index for the button
        >
          {selected.name}
          <svg
            className="-mr-1 ml-2 h-5 w-5"
            xmlns="http://www.w3.org/2000/svg"
            viewBox="0 0 20 20"
            fill="currentColor"
            aria-hidden="true"
          >
            <path
              fillRule="evenodd"
              d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 111.414 1.414l-4 4a1 1 01-1.414 0l-4-4a1 1 0 010-1.414z"
              clipRule="evenodd"
            />
          </svg>
        </button>
      </div>
      {isOpen && (
        <div
          className="origin-top-left absolute left-0 mt-2 w-56 rounded-md shadow-lg bg-white ring-1 ring-black ring-opacity-5"
          style={{ zIndex: 10 }}
        >
          <div
            className="py-1"
            role="menu"
            aria-orientation="vertical"
            aria-labelledby="options-menu"
          >
            {portfolios.map((portfolio) => (
              <button
                key={portfolio.code}
                onClick={() => handleSelect(portfolio)}
                className="block px-4 py-2 text-sm text-gray-700 hover:bg-gray-100 w-full text-left"
                role="menuitem"
              >
                {portfolio.name}
              </button>
            ))}
          </div>
        </div>
      )}
    </div>
  )
}
