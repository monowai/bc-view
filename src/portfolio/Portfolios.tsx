import React from "react";
import { Link } from "react-router-dom";
import { useHistory } from "react-router";
import { usePortfolios } from "./hooks";
import { isDone } from "../types/typeUtils";
import { ShowError } from "../errors/ShowError";

export function Portfolios(): React.ReactElement {
  const portfolioResults = usePortfolios();
  const history = useHistory();
  if (!isDone(portfolioResults)) {
    return (
      <div id="root" data-testid="loading">
        Loading...
      </div>
    );
  }
  if (portfolioResults.error) {
    return <ShowError error={portfolioResults.error} />;
  }
  const portfolios = portfolioResults.data;
  if (portfolios.length > 0) {
    return (
      <div>
        <nav className="container has-background-grey-lighter">
          <div className="column is-left">
            <button
              className="navbar-item button is-link is-small"
              onClick={() => {
                return history.push("/portfolios/new");
              }}
            >
              Create
            </button>
          </div>
        </nav>
        <div className="page-box is-primary has-background-light">
          <div className="container">
            <table className={"table is-striped is-hoverable"}>
              <thead>
                <tr>
                  <th>Code</th>
                  <th>Name</th>
                  <th>Report</th>
                  <th>Base</th>
                  <th>Actions</th>
                </tr>
              </thead>
              <tbody>
                {portfolios.map((portfolio) => (
                  <tr key={portfolio.id}>
                    <td>
                      <Link to={`/holdings/${portfolio.code}`}>{portfolio.code}</Link>
                    </td>
                    <td>{portfolio.name}</td>
                    <td>
                      {portfolio.currency.symbol}
                      {portfolio.currency.code}
                    </td>
                    <td>
                      {portfolio.base.symbol}
                      {portfolio.base.code}
                    </td>
                    <td>
                      <Link className="far fa-edit" to={`/portfolios/${portfolio.id}`} />
                      <span> </span>
                      <Link
                        className="far fa-trash-alt has-padding-left-6"
                        to={`/portfolios/${portfolio.id}/delete`}
                      />
                    </td>
                  </tr>
                ))}
              </tbody>
            </table>
          </div>
        </div>
      </div>
    );
  }
  return (
    <div id="root">
      You have no portfolios - <Link to={"/portfolios/new"}>create one?</Link>
    </div>
  );
}

export default Portfolios;
