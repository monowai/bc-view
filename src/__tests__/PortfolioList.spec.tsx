import React from "react";
import { cleanup, render, screen, waitForElementToBeRemoved } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import nock from "nock";
import Portfolios from "../portfolio/Portfolios";
import { MemoryRouter } from "react-router";
//https://kentcdodds.com/blog/common-mistakes-with-react-testing-library
afterEach(cleanup);
jest.mock("react-i18next", () => ({
  useTranslation: () => ({ t: (key) => key }),
}));

jest.mock("@react-keycloak/ssr", () => ({
  useKeycloak: () => ({
    initialized: true,
    keycloak: { token: "abc" },
  }),
}));

nock("http://localhost", {
  reqheaders: {
    authorization: "Bearer abc",
  },
})
  .get("/bff/portfolios")
  .replyWithFile(200, __dirname + "/__contracts__/portfolios.json", {
    "Access-Control-Allow-Origin": "*",
    "Content-type": "application/json",
  });

describe("<Portfolios />", () => {
  it("should match snapshot", async () => {
    const { container } = render(
      <MemoryRouter initialEntries={["/"]} keyLength={0}>
        <Portfolios />
      </MemoryRouter>
    );
    await screen.findByTestId("loading");
    await waitForElementToBeRemoved(() => screen.getByTestId("loading"));
    expect(nock.isDone());
    await screen.findByText("TEST");
    expect(container).toMatchSnapshot();
  });
});
