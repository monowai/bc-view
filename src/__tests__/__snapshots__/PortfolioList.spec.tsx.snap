// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Portfolios /> should match snapshot 1`] = `
<div>
  <div>
    <nav
      class="container has-background-grey-lighter"
    >
      <div
        class="column is-left"
      >
        <button
          class="navbar-item button is-link is-small"
        >
          Create
        </button>
      </div>
    </nav>
    <div
      class="page-box is-primary has-background-light"
    >
      <div
        class="container"
      >
        <table
          class="table is-striped is-hoverable"
        >
          <thead>
            <tr>
              <th>
                Code
              </th>
              <th>
                Name
              </th>
              <th>
                Report
              </th>
              <th>
                Base
              </th>
              <th>
                Actions
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                <a
                  href="/holdings/TEST"
                >
                  TEST
                </a>
              </td>
              <td>
                Test Portfolio
              </td>
              <td>
                â‚¬
                EUR
              </td>
              <td>
                $
                USD
              </td>
              <td>
                <a
                  class="far fa-edit"
                  href="/portfolios/Kdf3jfUfSYqXV_S4Myc2wA"
                />
                <span>
                   
                </span>
                <a
                  class="far fa-trash-alt has-padding-left-6"
                  href="/portfolios/Kdf3jfUfSYqXV_S4Myc2wA/delete"
                />
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
